resources/windows/wix/bootsnap.vbs

```diff
@@ -0,0 +1,197 @@
+  Dim BOLT_BASEDIR
+  Dim RUBY_DIR
+  Dim SSL_CERT_FILE
+  Dim SSL_CERT_DIR
+
+  BOLT_BASEDIR = fso.GetFolder(InstallDir).ShortPath
+
+  Log "BOLT_BASEDIR is " + BOLT_BASEDIR, false
+  RUBY_DIR = BOLT_BASEDIR + "\lib\ruby\"
+  SSL_CERT_FILE = BOLT_BASEDIR + "\ssl\cert.pem"
+  SSL_CERT_DIR = BOLT_BASEDIR + "\ssl\certs"
+
+  Log "BOLT_BASEDIR is " + BOLT_BASEDIR, false
+  Log "RUBY_DIR is " + RUBY_DIR, false
+  Log "SSL_CERT_FILE is " + SSL_CERT_FILE, false
+  Log "SSL_CERT_DIR is " + SSL_CERT_DIR, false
+
+  Dim RubyPath : RubyPath = BOLT_BASEDIR + "\bin\ruby.exe"
+  Dim BundlePath : BundlePath = BOLT_BASEDIR + "\bin\bundle"
+  Dim ProcessEnv : Set ProcessEnv = wshShell.Environment( "PROCESS" )
+
+  Log "Creating process level environment variables...", false
+  ProcessEnv("BOLT_BASEDIR") = BOLT_BASEDIR
+  ProcessEnv("RUBY_DIR") = RUBY_DIR
+  ProcessEnv("SSL_CERT_FILE") = SSL_CERT_FILE
+  ProcessEnv("SSL_CERT_DIR") = SSL_CERT_DIR
+  ProcessEnv("PDK_DEBUG") = "True"
+  Dim NewEnvPath : NewEnvPath = BOLT_BASEDIR + "\bin;" + ProcessEnv("PATH")
+  ProcessEnv("PATH") = NewEnvPath
+
+  ' Note Returning values back to the MSI Engine only works with Binary type Custom Actions
+  ' if ExecuteCommand("""" & RubyPath & """ -S -- """ & ExtractScript & """") Then
+  If ExecuteCommand(command) Then
+    Log "Completed with success", false
+    RunRubyCommand = IDOK
+  Else
+    Log "Completed with error", true
+    RunRubyCommand = IDABORT
+  End If
+End Function
+
+' Test whether a Folder is writeable by the current user by trying to write to
+' it and catching the error if it can't.
+Function FolderWriteable(FolderPath)
+  Dim FilePath
+  FilePath = fso.BuildPath(FolderPath, "TestWriteAccess.tmp")
+
+  ' Try to create a tmp file, and if that doesn't exit 0 return false
+  fso.CreateTextFile(FilePath).Write ("Test Folder Access")
+
+  ' Visual Basic Script doesn't have a way to 'catch' errors, but does have a
+  ' global Err object that contains the result of the last operation
+  ' performed. Check to see if that is 0, and return False if it's not.
+  ' https://stackoverflow.com/a/157785/2708055
+  If Err.Number <> 0 Then
+    FolderWriteable = False
+  Else
+    ' Cleanup temp file
+    fso.DeleteFile(FilePath)
+    ' If this hasn't errored, the folder is writeable
+    FolderWriteable = True
+  End If
+End Function
+
+Function RunBootsnap()
+  Dim BOLT_BASEDIR
+  Dim BOLT_PROGRAMDATA
+  Dim cmd
+  BOLT_BASEDIR = fso.GetFolder(InstallDir).ShortPath
+  BOLT_PROGRAMDATA = wshShell.ExpandEnvironmentStrings("%PROGRAMDATA%\PuppetLabs\Bolt")
+
+  ' If the ProgramData folder for Bolt does not exist create it
+  If Not(fso.FolderExists(BOLT_PROGRAMDATA)) Then
+    Log "Creating Bolt ProgramData folder", false
+    ' Create the Folders, one at a time
+    fso.CreateFolder(wshShell.ExpandEnvironmentStrings("%PROGRAMDATA%\PuppetLabs"))
+    fso.CreateFolder(BOLT_PROGRAMDATA)
+  End If
```

Comment:
Instead of detecting and creating here, we can add it to the WiX files like in the Puppet Agent project: https://github.com/puppetlabs/puppet-agent/blob/main/resources/windows/wix/appdatafiles.wxs. The puppet dir is created here: https://github.com/puppetlabs/puppet-agent/blob/main/resources/windows/wix/appdatafiles.wxs#L27-L31.

If we create the dir in the WiX, then the VBS can confidently write to the directory without worry about permissions or existence.

Reply:
What happens in that case if the directory already exists and isn't writeable? Or is that extremely unlikely?

Reply:
Ok, I feel like I may be duplicating some work here (like do we need the Wix and the `proj.directory` in the component?), but this at least has the intended behavior of setting up the directory!

Reply:
We need to create directories in WiX, so the MSI will handle removing them on uninstall. I don't think we need to put anything in vanagon components, because vanagon understands how to run WiX heat and candle correctly.